#162170722 Article search functionality

Johnsonojo opened the pull request:
#### What does this PR do?
It searches for articles by `tags`, `author's name`, and `keyword` 

 
#### Description of Task to be completed?
has the following endpoints working
* `GET /api/search/author/:author`
* `GET /api/search/article/:keyword`
* `GET /api/search/tag/:tagname`


#### How should this be manually tested?
- clone the repo
- run `npm install`
- change directory to the project folder
- run `npm run dev:start`
- use postman to navigate to any of the endpoints above to search for articles after posting 


#### Any background context you want to provide?
N/A


#### What are the relevant pivotal tracker stories? (If applicable)
[#162170722](https://www.pivotaltracker.com/story/show/162170722)


#### Screenshots (If applicable)
<img width="1280" alt="search functionality" src="https://user-images.githubusercontent.com/32887459/50006930-97722500-ffaf-11e8-9720-ede993f42669.png">

okoroemeka reviewed (approved):
Nice work @Johnsonojo

marcdomain commented on src/controllers/searchController.js:
```diff
@@ -0,0 +1,132 @@
+          }
+        ]
+      });
+      if (articles === null || articles.length === 0) {
```
Please what difference would it make if you had used `articles.length === 0` alone versus using `articles === null || articles.length === 0` ?

Johnsonojo replied:
Hi @marcdomain, nice observation. On second thought I think there is a need to refactor that aspect. There isn't much of a difference.

Johnsonojo replied:
@marcdomain, I have refactored it.

marcdomain commented on src/controllers/searchController.js:
```diff
@@ -0,0 +1,132 @@
+        attributes: ['id', 'username', 'image', 'email', 'image']
+      });
+      if (author) {
+        if (author.dataValues.articles.length < 1) {
```
what happens when `author.dataValues.articles.length === 0` versus `author.dataValues.articles.length < 1`?

Johnsonojo replied:
@marcdomain thanks for your observation. `author.dataValues.articles.length === 0` and `author.dataValues.articles.length < 1` returns the same result which is `true`. They are practically the same thing.

oluwajuwon replied:
@marcdomain I totally agree with you, it can't return anything that's '-1', but it's practically the same thing anyways

marcdomain commented on src/controllers/searchController.js:
```diff
@@ -0,0 +1,132 @@
+        ],
+        attributes: ['slug', 'title', 'description', 'body', 'imgUrl', 'readtime']
+      });
+      if (articles.length < 1) {
```
what happens when `articles.length === 0` versus `articles.length < 1`?

Johnsonojo replied:
same as above

frostyblok replied:
Hi Marcus, was the explanation clear enough?

marcdomain replied:
I think `articles.length < 1` is not necessary. Can the length here possibly be negative? Why not explicitly use `=== 0` or `!articles.length` to make it look elegant

Johnsonojo replied:
Hi Marcus, you're right. You can't have a negative length for the articles. But then `articles.length < 1` is less elaborate compared to `articles.length === 0`. I have refactored it though.

marcdomain reviewed (commented):
Please attend to the comments I left on `lines 36, 82, and 118` of `searchController.js`.
Also check if you can test for more lines in your codebase. The drop of `-0.2%` indicates that there are relevant tests that were ignored.

oluwajuwon commented on src/controllers/searchController.js:
```diff
@@ -0,0 +1,132 @@
+              title: { $iLike: `%${keyword}%` }
+            },
+            {
+              description: { $iLike: `%${keyword}%` }
```
@Johnsonojo nice idea here checking keywords in both the description and the title 👍

Johnsonojo replied:
Thanks. But @frostyblok actually implemented that.

akhilome commented on src/controllers/searchController.js:
```diff
@@ -0,0 +1,132 @@
+ */
+class SearchController {
+  /**
+   * @description - This method is responsible for creating new articles
```
Responsible for creating new articles?

Johnsonojo replied:
Nice observation Kizito. It has been fixed.

akhilome commented on src/controllers/searchController.js:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  /**
+   * @description - This method is responsible for creating new articles
```
Here too. This description is out of place.

Johnsonojo replied:
Fixed also.

akhilome reviewed (commented):
Nice job, guys (@Johnsonojo, @frostyblok)!

Please go over the few nitpicks I left below. 👇

Johnsonojo committed:
[feature #162170722] Adds search functionality and custom filtering
- User should be able to search article by authors name
- User should be able to search article by keyword and article title
- User should be able to search article by tags

Created the following endpoints:
- /api/search/author/:author
- /api/search/article/:keyword
- /api/search/tag/:tagname

[Finishes #162170722]
